### PR TITLE
Implement text draw using D2D WICBitmap render target

### PR DESCRIPTION
**Pull request by mnithish**

### CODE REVIEW

**Comment by mnithish** [comment]
@aballway is added to the review. #Closed

**Comment by mnithish** [comment]
@ms-jihua is added to the review. #Closed

**Comment by mnithish** [comment]
@rajsesh-msft is added to the review. #Closed

File: Frameworks/CoreGraphics/CGContextCairo.mm [hunk 1]
```diff
@@ -1977,3 +1985,62 @@
...
+    float height = _imgDest->Backing()->Height();
+
+    ComPtr<IWICImagingFactory> wicFactory;
+    CoCreateInstance(CLSID_WICImagingFactory, NULL, CLSCTX_INPROC_SERVER, __uuidof(IWICImagingFactory), &wicFactory);
+    THROW_NS_IF_NULL(E_UNEXPECTED, wicFactory);
+
+    ComPtr<IWICBitmap> wicBitmap;
+    THROW_IF_FAILED(
+        wicFactory->CreateBitmapFromMemory(width, height, GUID_WICPixelFormat32bppPBGRA, stride, height * stride, buffer, &wicBitmap));
+    THROW_NS_IF_NULL(E_UNEXPECTED, wicBitmap);
```
Comment (aballway) [inline Frameworks/CoreGraphics/CGContextCairo.mm]: >THROW_NS_IF_NULL [](start = 4, length = 16)

Will we not catch this in the THROW_IF_FAILED?

File: Frameworks/include/CoreTextInternal.h [hunk 2]
```diff
@@ -25,6 +25,14 @@
 #include <COMIncludes_End.h>
 #import <vector>
 
+template <typename T>
+inline void SafeRelease(T** p) {
```
Comment (aballway) [inline Frameworks/include/CoreTextInternal.h]: >SafeRelease [](start = 12, length = 11)

nit: should this be prefixed by __ ?

**Comment by aballway** [comment]
:shipit:

File: Frameworks/CoreGraphics/CGContextCairo.mm (hunk 1 above)
Comment (mnithish) [inline Frameworks/CoreGraphics/CGContextCairo.mm]: I guess I got the wrong file version checked in. I should cover CoCreateInstance with THROW_IF_FAILED

---
In reply to: [78452483](https://github.com/Microsoft/WinObjC/pull/999#discussion_r78452483) [](ancestors = 78452483)

File: Frameworks/include/CoreTextInternal.h (hunk 2 above)
Comment (mnithish) [inline Frameworks/include/CoreTextInternal.h]: Renamed it _SafeRelease

---
In reply to: [78454639](https://github.com/Microsoft/WinObjC/pull/999#discussion_r78454639) [](ancestors = 78454639)

File: Frameworks/CoreGraphics/CGContextCairo.mm [hunk 3]
```diff
@@ -1977,3 +1985,62 @@
...
+ * Today we create a IWICBitmap with the underlying XAML Writable Bitmap raw bytes and rasterize to it before copying it back to the
+ * original buffer.
+ * TODO::
+ *    1. This method today does not handle DPI scaling. Investigate what needs to be one here to support it
+ *    2. There is double copy happening here - first when the IWICBitmap is created and later the rasterized image is copied back
+ *       Implementing our own IWICBitmap interface should fix this issue.
+ *
+ * @parameter glyphRun DWRITE_GLYPH_RUN object to render
+ */
+void CGContextCairo::CGContextDrawGlyphRun(const DWRITE_GLYPH_RUN* glyphRun) {
```
Comment (ms-jihua) [inline Frameworks/CoreGraphics/CGContextCairo.mm]: nit: this function body could use a lot more comments - I can grasp the surface of what's happening, but I get the feeling that there's a lot of subtle stuff that someone else wouldn't pick up on first-time read?

File: Frameworks/CoreText/DWriteWrapper.mm [hunk 4]
```diff
@@ -47,12 +47,56 @@
     return s_dwriteFactory;
 }
 
+template <typename TElement>
+bool __CloneArray(_In_reads_opt_(count) TElement const* source,
```
Comment (ms-jihua) [inline Frameworks/CoreText/DWriteWrapper.mm]: This seems like the sort of thing that will have an STL solution already...

does std::copy suit your needs?

Comment (mnithish) [inline Frameworks/CoreText/DWriteWrapper.mm]: stl::copy only supports iterator based copy if I am not wrong?

---
In reply to: [78468942](https://github.com/Microsoft/WinObjC/pull/999#discussion_r78468942) [](ancestors = 78468942)

File: Frameworks/CoreGraphics/CGContextCairo.mm (hunk 3 above)
Comment (mnithish) [inline Frameworks/CoreGraphics/CGContextCairo.mm]: Sure will add

---
In reply to: [78468400](https://github.com/Microsoft/WinObjC/pull/999#discussion_r78468400) [](ancestors = 78468400)

File: Frameworks/CoreText/DWriteWrapper.mm (hunk 4 above)
Comment (aballway) [inline Frameworks/CoreText/DWriteWrapper.mm]: raw pointers are valid iterators for stl

---
In reply to: [78621358](https://github.com/Microsoft/WinObjC/pull/999#discussion_r78621358) [](ancestors = 78621358,78468942)

Comment (mnithish) [inline Frameworks/CoreText/DWriteWrapper.mm]: duh

---
In reply to: [78623064](https://github.com/Microsoft/WinObjC/pull/999#discussion_r78623064) [](ancestors = 78623064,78621358,78468942)

File: Frameworks/CoreGraphics/CGContextCairo.mm [hunk 5]
```diff
@@ -1977,3 +1985,66 @@
...
+    THROW_IF_FAILED(d2dFactory->CreateWicBitmapRenderTarget(wicBitmap.Get(), D2D1::RenderTargetProperties(), &renderTarget));
+
+    // Set the brush color to the current values from the context.
+    D2D1::ColorF brushColor =
+        D2D1::ColorF(curState->curFillColor.r, curState->curFillColor.g, curState->curFillColor.b, curState->curFillColor.a);
+
+    renderTarget->BeginDraw();
+
+    // Apply the required transformations as set in the context.
+    CGAffineTransform transform = CGAffineTransformConcat(curState->curTransform, curState->curTextMatrix);
```
Comment (ms-jihua) [inline Frameworks/CoreGraphics/CGContextCairo.mm]: >curState->curTransform [](start = 58, length = 22)

as discussed yesterday, should instead use a matrix with -b and -c in place of b and c

**Comment by ms-jihua** [comment]
:shipit:

File: Frameworks/CoreGraphics/CGContextCairo.mm (hunk 5 above)
Comment (mnithish) [inline Frameworks/CoreGraphics/CGContextCairo.mm]: Done. I also handled the wrong scaling we discussed yesterday

---
In reply to: [79202522](https://github.com/Microsoft/WinObjC/pull/999#discussion_r79202522) [](ancestors = 79202522)

**Pull request merged**